README.txt
Wx::GLCanvas is a wrapper for wxWindows' OpenGL canvas.

Copyright (c) 2003 Mattia Barbon.
This package is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

The files msw_myglcanvas.h, myglcanvas.h and glcanvas.cpp are
Copyright (c) Julian Smart, and are distributed under
the wxWindows licence (OSI certified, http://www.opensource.org/)

INSTALLATION:

Build and install wxPerl, then

perl Makefile.PL
make
make install



Changes
Revision history for Perl extension Wx::GLCanvas.

0.08  Thu Aug  6 20:12:12 CEST 2009
	- Add attribute list parameter to wxGLCanvas constructor
	  and wrap WX_GL_* constants.

0.07  Sun Jun 29 21:28:45 CEST 2008
	- Try to fail gracefully if wxPerl has not been
	  installed yet.

0.06  Sat Jun 23 22:03:29 CEST 2007
	- Compatibility with wxWidgets 2.9.x.

0.05  Sat Nov 11 17:14:03 CET 2006
	- Compilation fixes for wxWidgets 2.7.1 and 2.7.2.

0.04  Sun Aug 27 17:38:24 CEST 2006
	- Add overloaded Wx::GLCanvas constructors.
	- Make the cube rotate and have shiny colors.
	- Use the wxDemo framework for the example.

0.03  Wed Jan  5 07:43:03 CET 2005
	- Various fixes for wxWidgets 2.5.3 under Windows.

0.02  Fri Sep 12 23:35:08 CEST 2003
	- Compatibility with wxWindows 2.5.x
	- Don't depend on OpenGL. Any module providing access to OpenGL
	  API can be used together with Wx::GLCanvas.

0.01  Tue Aug 26 16:44:26 CEST 2003
	- First version.


Makefile.PL
#############################################################################
## Name:        Makefile.PL
## Purpose:     Makefile
## Author:      Mattia Barbon
## Modified by:
## Created:     26/07/2003
## RCS-ID:      $Id$
## Copyright:   (c) 2003, 2005, 2007-2008 Mattia Barbon
## Licence:     This program is free software; you can redistribute it and/or
##              modify it under the same terms as Perl itself
#############################################################################

use strict;
use lib 'inc';
use MyMakeMaker;

my %requires =
  ( 'Wx'                   => '0.57',
    'Wx::build::MakeMaker' => '0.16',
    );

eval {
    require ExtUtils::MY_Metafile;
    ExtUtils::MY_Metafile->import;

    my_metafile
      ( { license            => 'perl',
          dynamic_config     => 1,
          requires           => \%requires,
          build_requires     => {},
          configure_requires => { 'Wx::build::MakeMaker' => '0.16',
                                  },
          } );
};

wxWriteMakefile( NAME          => 'Wx::GLCanvas',
                 ABSTRACT_FROM => 'lib/Wx/GLCanvas.pm',
                 AUTHOR        => 'Mattia Barbon <mbarbon@cpan.org>',
                 VERSION_FROM  => 'lib/Wx/GLCanvas.pm',
                 LIBS          => ( $^O eq 'MSWin32' ? '-lopengl32' : '' ),
                 WX_CORE_LIB   => 'gl core base',
                 WX_OVERLOAD   => { header  => 'cpp/ovl_const.h',
                                    source  => 'cpp/ovl_const.cpp',
                                    },
                 PREREQ_PM     => \%requires,
                 LICENSE       => 'perl',
                 );


lib/Wx/GLCanvas.pm
#############################################################################
## Name:        lib/Wx/GLCanvas.pm
## Purpose:     loader for Wx::GLCanvas.pm
## Author:      Mattia Barbon
## Modified by:
## Created:     26/07/2003
## RCS-ID:      $Id$
## Copyright:   (c) 2003, 2005, 2007-2009 Mattia Barbon
## Licence:     This program is free software; you can redistribute it and/or
##              modify it under the same terms as Perl itself
#############################################################################

package Wx::GLCanvas;

use strict;
use Wx;
use base 'Wx::ScrolledWindow';

require Exporter; *import = \&Exporter::import;
our @EXPORT_OK =
  ( qw(WX_GL_RGBA WX_GL_BUFFER_SIZE WX_GL_LEVEL WX_GL_DOUBLEBUFFER
       WX_GL_STEREO WX_GL_AUX_BUFFERS WX_GL_MIN_RED WX_GL_MIN_GREEN
       WX_GL_MIN_BLUE WX_GL_MIN_ALPHA WX_GL_DEPTH_SIZE WX_GL_STENCIL_SIZE
       WX_GL_MIN_ACCUM_RED WX_GL_MIN_ACCUM_GREEN WX_GL_MIN_ACCUM_BLUE
       WX_GL_MIN_ACCUM_ALPHA) );
our %EXPORT_TAGS =
  ( all        => \@EXPORT_OK,
    everything => \@EXPORT_OK,
    );

$Wx::GLCanvas::VERSION = '0.09';

Wx::load_dll( 'gl' );
Wx::wx_boot( 'Wx::GLCanvas', $Wx::GLCanvas::VERSION );

our $AUTOLOAD;
sub AUTOLOAD {
  ( my $constname = $AUTOLOAD ) =~ s<^.*::>{};
  return if $constname eq 'DESTROY';
  my $val = constant( $constname, 0 );

  if( $! != 0 ) {
# re-add this if need support for autosplitted subroutines
#    $AutoLoader::AUTOLOAD = $AUTOLOAD;
#    goto &AutoLoader::AUTOLOAD;
    Wx::_croak( "Error while autoloading '$AUTOLOAD'" );
  }

  eval "sub $AUTOLOAD() { $val }";
  goto &$AUTOLOAD;
}

1;

__END__

=head1 NAME

Wx::GLCanvas - interface to wxWidgets' OpenGL canvas

=head1 SYNOPSIS

    use OpenGL; # or any other module providing OpenGL API
    use Wx::GLCanvas;

=head1 DESCRIPTION

The documentation for this module is included in the main
wxPerl distribution (wxGLCanvas).

=head1 AUTHOR

Mattia Barbon <mbarbon@cpan.org>

=head1 LICENSE

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut

# local variables:
# mode: cperl
# end:


lib/Wx/DemoModules/wxGLCanvas.pm
#############################################################################
## Name:        lib/Wx/DemoModules/wxGLCanvas.pm
## Purpose:     wxPerl demo helper for Wx::GLCanvas
## Author:      Mattia Barbon
## Modified by:
## Created:     26/07/2003
## RCS-ID:      $Id$
## Copyright:   (c) 2000, 2006, 2008 Mattia Barbon
## Licence:     This program is free software; you can redistribute it and/or
##              modify it under the same terms as Perl itself
#############################################################################

package Wx::DemoModules::wxGLCanvas;

use strict;

use Wx::Event qw(EVT_PAINT EVT_SIZE EVT_ERASE_BACKGROUND EVT_IDLE EVT_TIMER);
# must load OpenGL *before* Wx::GLCanvas
use OpenGL qw(:glconstants :glfunctions);
use base qw(Wx::GLCanvas Class::Accessor::Fast);
use Wx::GLCanvas qw(:all);

__PACKAGE__->mk_accessors( qw(timer x_rot y_rot dirty init) );

sub new {
    my( $class, $parent ) = @_;
    my $self = $class->SUPER::new( $parent );
#     my $self = $class->SUPER::new( $parent, -1, [-1, -1], [-1, -1], 0,
#                                    "GLCanvas",
#                                    [WX_GL_RGBA, WX_GL_DOUBLEBUFFER, 0] );

    my $timer = $self->timer( Wx::Timer->new( $self ) );
    $timer->Start( 50 );

    $self->x_rot( 0 );
    $self->y_rot( 0 );

    EVT_PAINT( $self,
               sub {
                   my $dc = Wx::PaintDC->new( $self );
                   $self->Render( $dc );
               } );
    EVT_SIZE( $self, sub { $self->dirty( 1 ) } );
    EVT_IDLE( $self, sub {
                  return unless $self->dirty;
                  $self->Resize( $self->GetSizeWH );
                  $self->Refresh;
              } );
    EVT_TIMER( $self, -1, sub {
                   my( $self, $e ) = @_;

                   $self->x_rot( $self->x_rot - 1 );
                   $self->y_rot( $self->y_rot + 2 );

                   $self->dirty( 1 );
                   Wx::WakeUpIdle;
               } );

    return $self;
}

sub GetContext {
    my( $self ) = @_;

    if( Wx::wxVERSION >= 2.009 ) {
        return $self->{context} ||= Wx::GLContext->new( $self );
    } else {
        return $self->SUPER::GetContext;
    }
}

sub SetCurrent {
    my( $self, $context ) = @_;

    if( Wx::wxVERSION >= 2.009 ) {
        return $self->SUPER::SetCurrent( $context );
    } else {
        return $self->SUPER::SetCurrent;
    }
}

sub Resize {
    my( $self, $x, $y ) = @_;

    return unless $self->GetContext;
    $self->dirty( 0 );

    $self->SetCurrent( $self->GetContext );
    glViewport( 0, 0, $x, $y );

    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    my_gluPerspective( 45, $x/$y, .5, 100 );

    glMatrixMode(GL_MODELVIEW);
}

use Math::Trig;

sub my_gluPerspective {
    my( $fov, $ratio, $near, $far ) = @_;

    my $top = tan(deg2rad($fov)*0.5) * $near;
    my $bottom = -$top;
    my $left = $ratio * $bottom;
    my $right = $ratio * $top;

    glFrustum( $left, $right, $bottom, $top, $near, $far );
}

sub DESTROY {
    my( $self ) = @_;

    $self->timer->Stop;
    $self->timer( undef );
}

sub tags { [ 'windows/glcanvas' => 'wxGLCanvas' ] }

package Wx::DemoModules::wxGLCanvas::Cube;

use strict;

# must load OpenGL *before* Wx::GLCanvas
use OpenGL qw(:glconstants :glfunctions);
use base qw(Wx::DemoModules::wxGLCanvas);

sub cube {
    my( @v ) = ( [ 1, 1, 1 ], [ -1, 1, 1 ],
                 [ -1, -1, 1 ], [ 1, -1, 1 ],
                 [ 1, 1, -1 ], [ -1, 1, -1 ],
                 [ -1, -1, -1 ], [ 1, -1, -1 ] );
    my( @c ) = ( [ 1, 1, 0 ], [ 1, 0, 1 ],
                 [ 0, 1, 1 ], [ 1, 1, 1 ],
                 [ 0, 0, 1 ], [ 0, 1, 0 ],
                 [ 1, 0, 1 ], [ 1, 1, 0 ] );
    my( @s ) = ( [ 0, 1, 2, 3 ], [ 4, 5, 6, 7 ],
                 [ 0, 1, 5, 4 ], [ 2, 3, 7, 6 ],
                 [ 1, 2, 6, 5 ], [ 0, 3, 7, 4 ] );

    for my $i ( 0 .. 5 ) {
        my $s = $s[$i];
        glBegin(GL_QUADS);
        foreach my $j ( @$s ) {
            glColor3f( @{$c[$j]} );
            glVertex3f( @{$v[$j]} );
        }
        glEnd();
    }
}

sub InitGL {
    my $self = shift;

    return if $self->init;
    return unless $self->GetContext;
    $self->init( 1 );

    glDisable( GL_LIGHTING );
    glDepthFunc( GL_LESS );
    glEnable( GL_DEPTH_TEST );
}

sub Render {
    my( $self, $dc ) = @_;

    return unless $self->GetContext;
    $self->SetCurrent( $self->GetContext );
    $self->InitGL;

    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    glPushMatrix();
    glTranslatef( 0, 0, -5 );
    glRotatef( $self->x_rot, 1, 0, 0 );
    glRotatef( $self->y_rot, 0, 0, 1 );

    cube();

    glPopMatrix();
    glFlush();

    $self->SwapBuffers();
}

sub add_to_tags { qw(windows/glcanvas) }
sub title { 'Cube' }
sub file { __FILE__ }

package Wx::DemoModules::wxGLCanvas::Light;

use strict;

# must load OpenGL *before* Wx::GLCanvas
use OpenGL qw(:glconstants :glfunctions);
use base qw(Wx::DemoModules::wxGLCanvas);

sub plane_mat {
    glMaterialfv_p( GL_FRONT, GL_AMBIENT_AND_DIFFUSE, 0.2, 0.1, 0.8, 0.5 );
    glMaterialfv_p( GL_FRONT, GL_SPECULAR, 1, 1, 1, 1 );
    glMaterialfv_p( GL_FRONT, GL_SHININESS, 4 );
}

sub plane {
    my( $x, $y, $step ) = ( -2, -2, .5 );

    for my $i ( 0 .. 7 ) {
        $x += $step;
        $y = -2;
        for my $j ( 0 .. 7 ) {
            $y += $step;
            glBegin(GL_QUADS);
                glNormal3f(  0,         1,  0 );
                glVertex3f( $x,         -1, $y );
                glVertex3f( $x,         -1, $y + $step );
                glVertex3f( $x + $step, -1, $y + $step );
                glVertex3f( $x + $step, -1, $y );
            glEnd();
        }
    }
}

sub icosaedron_mat {
    glMaterialfv_p( GL_FRONT, GL_AMBIENT_AND_DIFFUSE, 0.1, 0.4, 0.7, 1 );
    glMaterialfv_p( GL_FRONT, GL_SPECULAR, 1, 0, 0, 1 );
    glMaterialfv_p( GL_FRONT, GL_SHININESS, 5 );
    glMaterialfv_p( GL_FRONT, GL_EMISSION, 0.1, 0.2, 0.7, 0 );
}

sub icosahedron{
    # from the 'light' demo in OpenGL-0.54
    # from OpenGL Programming Guide page 56
    my $x = 0.525731112119133606;
    my $z = 0.850650808352039932;

    my $v=[
	[-$x,   0,  $z],
	[ $x,   0,  $z],
	[-$x,   0, -$z],
	[ $x,   0, -$z],
	[  0,  $z,  $x],
	[  0,  $z, -$x],
	[  0, -$z,  $x],
	[  0, -$z, -$x],
	[ $z,  $x,   0],
	[-$z,  $x,   0],
	[ $z, -$x,   0],
	[-$z, -$x,   0],
       ];
    my $t=[
	[0,4,1],  	[0, 9, 4],
    	[9, 5, 4],    	[4, 5, 8],
    	[4, 8, 1],    	[8, 10, 1],
    	[8, 3, 10],    	[5, 3, 8],
    	[5, 2, 3],    	[2, 7, 3],
    	[7, 10, 3],    	[7, 6, 10],
    	[7, 11, 6],    	[11, 0, 6],
    	[0, 1, 6],    	[6, 1, 10],
    	[9, 0, 11],    	[9, 11, 2],
    	[9, 2, 5],    	[7, 2, 11],
       ];
    for(my $i=0;$i<20;$i++) {
	glBegin(GL_POLYGON);
	    for(my $j=0;$j<3;$j++) {
		glNormal3f( $v->[$t->[$i][$j]][0],
				$v->[$t->[$i][$j]][1],
				$v->[$t->[$i][$j]][2]);
		glVertex3f( $v->[$t->[$i][$j]][0],
				$v->[$t->[$i][$j]][1],
				$v->[$t->[$i][$j]][2]);
	    }
	glEnd();
    }
}

sub InitGL {
    my $self = shift;

    return if $self->init;
    return unless $self->GetContext;
    $self->init( 1 );

    glLightfv_p( GL_LIGHT0, GL_AMBIENT, .5,.5,.5,1 );
    glLightfv_p( GL_LIGHT0, GL_DIFFUSE, .2,.2,.2,1 );
    glLightfv_p( GL_LIGHT0, GL_SPECULAR, 1,.5,.5,1 );
    glLightfv_p( GL_LIGHT0, GL_POSITION, -3, 3, -4, 1);

    glLightfv_p( GL_LIGHT1, GL_SPECULAR, 0, 1, 1, 1 );
    glLightfv_p( GL_LIGHT1, GL_POSITION, 0, 0, -6, 1 );

    glLightfv_p( GL_LIGHT2, GL_SPECULAR, 1, 1, 1, 1 );
    glLightfv_p( GL_LIGHT2, GL_POSITION, 2, -0.5, -4, 1 );

    glEnable( GL_LIGHTING );
    glEnable( GL_LIGHT0 );
    glEnable( GL_LIGHT1 );
    glEnable( GL_LIGHT2 );
    glDepthFunc( GL_LESS );
    glEnable( GL_DEPTH_TEST );
}

sub Render {
    my( $self, $dc ) = @_;

    return unless $self->GetContext;
    $self->SetCurrent( $self->GetContext );

    $self->InitGL;

    glClear( GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT );
    glPushMatrix();
    glTranslatef( 0, 0, -5 );

    plane_mat();
    plane();

    glRotatef( $self->x_rot, 1, 0, 0 );
    glRotatef( $self->y_rot, 0, 0, 1 );

    icosaedron_mat();
    icosahedron();

    glPopMatrix();
    glFlush();

    $self->SwapBuffers();
}

sub add_to_tags { qw(windows/glcanvas) }
sub title { 'Light' }
sub file { __FILE__ }

1;
